src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -854,8 +857,15 @@ private static JsonSerializerOptions GetOrCreateDefaultOptionsInstance()
         {
             var options = new JsonSerializerOptions
             {
-                TypeInfoResolver = DefaultJsonTypeInfoResolver.RootDefaultInstance(),
-                _isReadOnly = true
+                // Because're marking the default instance as read-only,
+                // we need to specify a resolver instance for the case where
+                // reflection is disabled by default: use one that returns null for all types.
+
+                TypeInfoResolver = AppContextSwitchHelper.IsDefaultReflectionDisabled
+                    ? JsonTypeInfoResolver.Combine()
```

Comment:
(nit) this looks weird, and isn't really readable. "Combine" what? Nothing?

```suggestion
                    ? new CombiningJsonTypeInfoResolver(Array.Empty<IJsonTypeInfoResolver>())
```

Might be a little better? Or even just making a simple `NoopJsonTypeInfoResolver` class that always `return null;`.

Reply:
The two forms are identical, since they both define an empty resolver chain. I considered creating a bespoke `NullJsonTypeInfoResolver` but I don't think it would be of much value unless we made the type public so that users can pattern match on it.